Source/JavaScriptCore/runtime/TemporalDuration.cpp

```diff
@@ -378,6 +378,92 @@ ISO8601::Duration TemporalDuration::add(JSGlobalObject* globalObject, JSValue ot
     return result;
 }
 
+// https://tc39.es/proposal-temporal/#sec-temporal-temporaldurationfrominternal
+ISO8601::Duration TemporalDuration::temporalDurationFromInternal(ISO8601::InternalDuration internalDuration,
+    TemporalUnit largestUnit)
+{
+    double days = 0;
+    double hours = 0;
+    double minutes = 0;
+    double seconds = 0;
+    Int128 milliseconds = 0;
+    Int128 microseconds = 0;
+
+    int32_t sign = internalDuration.timeDurationSign();
+    Int128 nanoseconds = absInt128(internalDuration.time());
+
+    if (largestUnit <= TemporalUnit::Day) {
+        microseconds = nanoseconds / 1000;
+        nanoseconds = nanoseconds % 1000;
+        milliseconds = microseconds / 1000;
+        microseconds = microseconds % 1000;
+        seconds = (double) (milliseconds / 1000);
+        milliseconds = milliseconds % 1000;
+        minutes = std::trunc(seconds / 60);
+        seconds = std::fmod(seconds, 60);
+        hours = std::trunc(minutes / 60);
+        minutes = std::fmod(minutes, 60);
+        days = std::trunc(hours / 24);
+        hours = std::fmod(hours, 24);
+    } else if (largestUnit == TemporalUnit::Hour) {
+        microseconds = nanoseconds / 1000;
+        nanoseconds = nanoseconds % 1000;
+        milliseconds = microseconds / 1000;
+        microseconds = microseconds % 1000;
+        seconds = (double) (milliseconds / 1000);
+        milliseconds = milliseconds % 1000;
+        minutes = std::trunc(seconds / 60);
+        seconds = std::fmod(seconds, 60);
+        hours = std::trunc(minutes / 60);
+        minutes = std::fmod(minutes, 60);
+    } else if (largestUnit == TemporalUnit::Minute) {
+        microseconds = nanoseconds / 1000;
+        nanoseconds = nanoseconds % 1000;
+        milliseconds = microseconds / 1000;
+        microseconds = microseconds % 1000;
+        seconds = (double) (milliseconds / 1000);
+        milliseconds = milliseconds % 1000;
+        minutes = std::trunc(seconds / 60);
+        seconds = std::fmod(seconds, 60);
+    } else if (largestUnit == TemporalUnit::Second) {
+        microseconds = nanoseconds / 1000;
+        nanoseconds = nanoseconds % 1000;
+        milliseconds = microseconds / 1000;
+        microseconds = microseconds % 1000;
+        seconds = (double) (milliseconds / 1000);
+        milliseconds = milliseconds % 1000;
+    } else if (largestUnit == TemporalUnit::Millisecond) {
+        microseconds = nanoseconds / 1000;
+        nanoseconds = nanoseconds % 1000;
+        milliseconds = microseconds / 1000;
+        microseconds = microseconds % 1000;
+    } else if (largestUnit == TemporalUnit::Microsecond) {
+        microseconds = (Int128) std::trunc(std::fma<double, unsigned, double>(
+            seconds, 6, std::trunc(((double) nanoseconds) / 1000)));
```

Comment:
I think 6 is probably meant to be 10⁶? If so, this should make more test262 tests pass (and if it doesn't, we should add coverage for this)

Reply:
Good catch! It turns out the code was correct because `seconds` is always 0. (I got there from following the code in the polyfill, but in the polyfill time durations are represented with seconds and subseconds separately, and in JSC it's a single nanoseconds value.) But the call to `std::fma` is unnecessary so I changed it in [c719f90](https://github.com/WebKit/WebKit/pull/45603/commits/c719f90b74999ab3f185aa9f6f54ad72d3749c37) to be similar to the other cases.